File: apps/web/lib/wake.test.ts
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { wakeReadiness, wakeSenders, wakeSentText, type WakeablePc } from './wake.js';

function pc(overrides: Partial<WakeablePc> & { id: string }): WakeablePc {
  return {
    name: overrides.id,
    status: 'offline',
    remoteAccessEnabled: true,
    capabilities: { wakeOnLanCapable: true, wakeAddressKnown: true, supportedCommands: ['power.wake'] },
    ...overrides,
  };
}

test('a PC is ready to wake only when it is offline, reachable by policy, and has a known armed adapter', () => {
  assert.equal(wakeReadiness(pc({ id: 'a' })).kind, 'ready');
  assert.equal(wakeReadiness(pc({ id: 'a', status: 'online' })).kind, 'online');
  assert.equal(wakeReadiness(pc({ id: 'a', remoteAccessEnabled: false })).kind, 'switched-off');
  assert.equal(
    wakeReadiness(pc({ id: 'a', capabilities: { wakeOnLanCapable: false, wakeAddressKnown: false, supportedCommands: [] } })).kind,
    'no-address',
  );
  assert.equal(wakeReadiness(pc({ id: 'a', capabilities: null })).kind, 'no-address');
  assert.equal(
    wakeReadiness(pc({ id: 'a', capabilities: { wakeOnLanCapable: false, wakeAddressKnown: true, supportedCommands: [] } })).kind,
    'not-armed',
  );
});

test('senders are the owner’s other PCs that are online, not switched off, and can send a wake', () => {
  const target = pc({ id: 'tower' });
  const senders = wakeSenders(target, [
    target,
    pc({ id: 'zeta', status: 'online' }),
    pc({ id: 'alpha', status: 'online' }),
    pc({ id: 'asleep' }),
    pc({ id: 'off', status: 'online', remoteAccessEnabled: false }),
    pc({ id: 'old-agent', status: 'online', capabilities: { wakeOnLanCapable: false, supportedCommands: ['power.action'] } }),
  ]);

  assert.deepEqual(senders.map((sender) => sender.id), ['alpha', 'zeta']);
});

test('the result says what was sent, never that the PC woke', () => {
  const text = wakeSentText('Laptop', 'Tower', { packetsSent: 6, networks: 1 });
  assert.match(text, /Laptop sent 6 wake packets for Tower across 1 local network\./);
  assert.doesNotMatch(text, /woke Tower|Tower woke|is awake/);
});

File: apps/android/app/src/test/java/app/amizhthan/wolf/api/WakeTest.kt
package app.amizhthan.wolf.api

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/** The web dashboard's wake rules, on the phone. */
class WakeTest {
    private fun pc(
        id: String,
        status: String = "offline",
        remoteAccessEnabled: Boolean = true,
        capabilities: PcCapabilitiesSummary? = PcCapabilitiesSummary(wakeOnLanCapable = true, wakeAddressKnown = true, supportedCommands = listOf("power.wake")),
    ) = PcSummary(id = id, name = id, status = status, registrationState = "active", remoteAccessEnabled = remoteAccessEnabled, capabilities = capabilities)

    @Test
    fun the_command_names_only_the_pc_to_wake() {
        assertEquals("""{"type":"power.wake","payload":{"targetPcId":"01J9ZQK7T0000000000000000A"}}""", Wake.command("01J9ZQK7T0000000000000000A").toString())
    }

    @Test
    fun readiness_follows_status_policy_and_what_the_pc_reported() {
        assertEquals(Wake.Readiness.READY, Wake.readiness(pc("a")))
        assertEquals(Wake.Readiness.ONLINE, Wake.readiness(pc("a", status = "online")))
        assertEquals(Wake.Readiness.SWITCHED_OFF, Wake.readiness(pc("a", remoteAccessEnabled = false)))
        assertEquals(Wake.Readiness.NO_ADDRESS, Wake.readiness(pc("a", capabilities = null)))
        assertEquals(Wake.Readiness.NO_ADDRESS, Wake.readiness(pc("a", capabilities = PcCapabilitiesSummary())))
        assertEquals(Wake.Readiness.NOT_ARMED, Wake.readiness(pc("a", capabilities = PcCapabilitiesSummary(wakeAddressKnown = true))))
    }

    @Test
    fun senders_are_other_online_pcs_that_can_send_a_wake() {
        val target = pc("tower")
        val senders = Wake.senders(
            target,
            listOf(
                target,
                pc("Zeta", status = "online"),
                pc("alpha", status = "online"),
                pc("asleep"),
                pc("off", status = "online", remoteAccessEnabled = false),
                pc("old", status = "online", capabilities = PcCapabilitiesSummary(supportedCommands = listOf("power.action"))),
            ),
        )
        assertEquals(listOf("alpha", "Zeta"), senders.map { it.id })
    }

    @Test
    fun the_result_says_what_was_sent_never_that_the_pc_woke() {
        val text = Wake.sentText("Laptop", "Tower", 6, 1)
        assertTrue(text.startsWith("Laptop sent 6 wake packets for Tower across 1 local network."))
        assertFalse(text.contains("Tower woke"))
    }

    @Test
    fun a_pc_list_from_before_capabilities_were_listed_still_reads() {
        val parsed = WolfJson.decodeFromString(PcList.serializer(), """{"pcs":[{"id":"a","name":"A","status":"offline","registrationState":"active"}]}""")
        assertNull(parsed.pcs.single().capabilities)
    }
}
